src/rlx_resolve.erl

```diff
@@ -0,0 +1,338 @@
+                          Type =:= load ;
+                          Type =:= none ->
+            maybe_with_type({Name, Vsn}, Type);
+        {Name, Vsn1, Type} when Type =:= permanent ;
+                                Type =:= transient ;
+                                Type =:= temporary ;
+                                Type =:= load ;
+                                Type =:= none ->
+            %% TODO: error if Vsn and Vsn1 don't match
+            maybe_with_type({Name, Vsn1}, Type);
+        {Name, IncludedApplications} when is_list(IncludedApplications) ->
+            %% TODO: error if Vsn and Vsn1 don't match
+            maybe_with_type({Name, Vsn, IncludedApplications}, Type0);
+        _ ->
+            %% TODO: throw a real error
+            error
+    end.
+
+%% keep a clean .rel file by only included non-defaults
+maybe_with_type(Tuple, permanent) ->
+    Tuple;
+maybe_with_type(Tuple, Type) ->
+    erlang:insert_element(3, Tuple, Type).
+
+-spec realize_erts(release()) -> release().
+realize_erts(Rel=#release{erts=undefined}) ->
+    Rel#release{erts=erlang:system_info(version)};
+realize_erts(Rel) ->
+    Rel.
+
+%%
+
+find(Key, []) ->
+    Key;
+find(Key, [Key | _]) ->
+    Key;
+find(Key, [Tuple | _]) when is_tuple(Tuple) ,
+                            element(1, Tuple) =:= Key->
+    Tuple;
+find(Key, [_ | Rest]) ->
+    find(Key, Rest).
+
+%% format errors
+
+-spec format_error(ErrorDetail::term()) -> iolist().
+format_error(no_goals_specified) ->
+    "No goals specified for this release ~n";
+format_error({release_erts_error, Dir}) ->
+    io_lib:format("Unable to find erts in ~s~n", [Dir]);
+format_error({no_release_name, Vsn}) ->
+    io_lib:format("A target release version was specified (~s) but no name", [Vsn]);
+format_error({invalid_release_info, Info}) ->
+    io_lib:format("Target release information is in an invalid format ~p", [Info]);
+format_error({multiple_release_names, RelA, RelB}) ->
+    io_lib:format("No default release name was specified and there are multiple "
+                 "releases in the config: ~s, ~s",
+                 [RelA, RelB]);
+format_error(no_releases_in_system) ->
+    "No releases have been specified in the system!";
+format_error({no_releases_for, RelName}) ->
+    io_lib:format("No releases exist in the system for ~s!", [RelName]);
+format_error({release_not_found, {RelName, RelVsn}}) ->
+    io_lib:format("No releases exist in the system for ~p:~s!", [RelName, RelVsn]);
+format_error({failed_solve, Error}) ->
+    io_lib:format("Failed to solve release:\n ~s",
+                  [rlx_depsolver:format_error({error, Error})]);
+format_error({release_error, Error}) ->
+    io_lib:format("Failed to resolve release:\n ~p~n", [Error]).
+
+%% legacy
+
+-spec find_default_release(rlx_state:t()) -> {ok, rlx_state:t()} | relx:error().
+find_default_release(State) ->
+    try
+        case rlx_state:default_configured_release(State) of
```

Comment:
should this be a `try <exp> of` ?

Reply:
Looks like it. This is code that was just moved, not touched besides moving so far.

Reply:
Or no, the `multiple_release` is actually thrown from the other functions in here and not from `rlx_state:default_configured_release`. So needs to be a `try` around the `case`.